core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java

```diff
@@ -224,27 +233,35 @@ DistributedTransaction begin(String txId, Isolation isolation) {
     Snapshot snapshot = new Snapshot(txId, isolation, tableMetadataManager, parallelExecutor);
     CrudHandler crud =
         new CrudHandler(
-            storage, snapshot, tableMetadataManager, isIncludeMetadataEnabled, parallelExecutor);
-    ConsensusCommit consensus =
+            storage,
+            snapshot,
+            tableMetadataManager,
+            isIncludeMetadataEnabled,
+            parallelExecutor,
+            readOnly);
+    DistributedTransaction transaction =
         new ConsensusCommit(crud, commit, recovery, mutationOperationChecker, groupCommitter);
-    getNamespace().ifPresent(consensus::withNamespace);
-    getTable().ifPresent(consensus::withTable);
-    return consensus;
+    if (readOnly) {
+      transaction = new ReadOnlyDistributedTransaction(transaction);
```

Comment:
Wrap the transaction object with `ReadOnlyDistributedTransaction` for read-only transactions.